### ISSUE TITLE
Extract needs to be refactored so that reweighting is in one place in the code

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#710]
Currently for extract, wind photons are typically reweighted  (for various scattering modes) to a maximum value in trans_phot and then de-weighted given a photon direction in extract.  This results in duplicated code, and code that is difficult to understand.  

The code needs to be refactored so that all the reweighting is carried out in extract.  This will make it easier to deal with comoving frames (#285)

**Comment by kslong** [comment]
@jhmatthews  I now have successfully moved the weighting out of trans_phot into the beginning of extract.  The problem I was having all came down to nnscat.  For the initial extract one uses the nnscat that is contained in the photons.  As far as I can determine this is always 1.  As photons go out of through the wind, a variable in trans_phot  nnscat is incremented, and this must be copied into the photon (or the one that is created from p and then sent to extract).  

Though I have produced the same result as dev, I do not understand what is happening here.  As we discussed yesterday, I don't see why the reweighting of the photon after a scatter, should depend on the number of previous scatters.  I would greatly appreciate an explanation for this, and I think some comment should be added to the code to explain the reason.  

I will likely go ahead and merge my changes for this into dev.  The regression tests all work, and give the same answers as previously

**Comment by kslong** [comment]
This change has been merged.  The only reason for not closing it is my question to @jhmatthews  above.

**Comment by jhmatthews** [comment]
So the basic changes look good here.

> Though I have produced the same result as dev, I do not understand what is happening here. As we discussed yesterday, I don't see why the reweighting of the photon after a scatter, should depend on the number of previous scatters. I would greatly appreciate an explanation for this, and I think some comment should be added to the code to explain the reason.

I haven't been able to track down the document I wrote about this, but I believe the logic here is as follows (we have discussed before, but putting here for completeness). 

Firstly, let's review the rejection method in anisowind:
* To choose a direction, we randomly sample directions and calculate the escape probability along that direction with index i, P(i,esc). 
* each time, we escape if P(i,esc)>z, where z is a random number between 0 and 1. 
* in some cases, the escape probability along every direction is small, so this rejection method is extremely inefficient (see #82). To deal with this, we normalise the rejection method by calculating P_norm, which is the escape probability along dvds_max. 
* we then set z-> z*P_norm, which ensures that the rejection method is always fairly efficient (i.e. a reasonable amount of the probability density space is filled with by the escape probability surface in theta, phi). 

That's all fine, but when we extract photons we need to know what fraction of the photons ecaped in a given direction so we can reweight things. The reweighting factor we want is 

f_reweight = P_esc,i / mean(P_esc,i) 

The statement is made in the code that mean(P_esc,i) = 1/nnscat. I think the point here is that, for an unnormalised rejection method, the number of tries in the rejection method is equal to total_probability/(the mean escape probability), and since the total probability is 1.0 the mean escape probability is 1/nnscat. In reality, we've normalised the rejection method, so nnscat will be lower, and we need an additional factor P_norm, so 

f_reweight = P_esc,i * (nnscat / P_norm). 

This is currently done in two stages (lines 135 and 343 in extract.c in Knox's refactored version). 

Note that really nnscat is only proportional to mean(P_esc,i), not equal to it, so we are only correct in a stochastic sense. One could improve this by actually calculating mean(P_esc,i) with a large number of sample rays in the spectral cycles, but it's probably not so bad. 

If you're happy with that explanation and this "stochastic" approach to getting mean(P_esci,i) Knox, we can close this.
